File: core/src/lin_algebra.c
/*
This source file is part of ndvar, which is released under the MIT license.
Github repository: https://github.com/AUN4GFD/ndvar
*/

/*
Linear algebra functions.
*/

#include <stdlib.h>
#include "ndvar.h"

int inv_gauss(double to_be_inverted[][NO_OF_CHOSEN_OBSERVATIONS], double inv[][NO_OF_CHOSEN_OBSERVATIONS])
{
	/*
	This function computes the inverse inv of the matrix to_be_inverted, using the Gauss scheme.
	CAUTION: in the process, to_be_inverted will be modified.
	*/
	
	// firstly, the inverse is initialized with the unity matrix
	for (int i = 0; i < NO_OF_CHOSEN_OBSERVATIONS; ++i)
	{
		inv[i][i] = 1;
	}
	
	/*
	Gaussian downwards
	------------------
	we will start to modify to_be_inverted now (misuse of name)
	*/
	double factor;
	for (int i = 0; i < NO_OF_CHOSEN_OBSERVATIONS - 1; ++i)
	{
		// dividing the line by to_be_inverted[i][i]
		factor = 1/to_be_inverted[i][i];
		for (int j = i; j < NO_OF_CHOSEN_OBSERVATIONS; ++j)
		{
			to_be_inverted[i][j] = factor*to_be_inverted[i][j];
		}
		for (int j = 0; j <= i; ++j)
		{
			inv[i][j] = factor*inv[i][j];
		}
		// loop over all the lines that are below the current line
		for (int j = i + 1; j < NO_OF_CHOSEN_OBSERVATIONS; ++j)
		{
			factor = -to_be_inverted[j][i];
			for (int k = i; k < NO_OF_CHOSEN_OBSERVATIONS; ++k)
			{
				to_be_inverted[j][k] = to_be_inverted[j][k] + factor*to_be_inverted[i][k];
			}
			for (int k = 0; k <= i; ++k)
			{
				inv[j][k] = inv[j][k] + factor*inv[i][k];
			}
		}
	}
	
	for (int j = 0; j < NO_OF_CHOSEN_OBSERVATIONS; ++j)
	{
		inv[NO_OF_CHOSEN_OBSERVATIONS - 1][j] = inv[NO_OF_CHOSEN_OBSERVATIONS - 1][j]/to_be_inverted[NO_OF_CHOSEN_OBSERVATIONS - 1][NO_OF_CHOSEN_OBSERVATIONS - 1];
	}
	to_be_inverted[NO_OF_CHOSEN_OBSERVATIONS - 1][NO_OF_CHOSEN_OBSERVATIONS - 1] = 1;
	
	/*
	Gaussian upwards
	----------------
	*/
	for (int i = NO_OF_CHOSEN_OBSERVATIONS - 1; i >= 1; --i)
	{
		for (int j = i - 1; j >= 0; --j)
		{
			for (int k = 0; k < NO_OF_CHOSEN_OBSERVATIONS; ++k)
			{
				inv[j][k] = inv[j][k] - to_be_inverted[j][i]*inv[i][k];
			}
		}
	}
	return 0;
}

int inv_lu(double to_be_inverted[][NO_OF_CHOSEN_OBSERVATIONS], double inv[][NO_OF_CHOSEN_OBSERVATIONS])
{
	/*
	This function computes the inverse inv of the matrix to_be_inverted, using the LU decomposition.
	CAUTION: in the process, to_be_inverted will be modified.
	*/
	double (*l_matrix)[NO_OF_CHOSEN_OBSERVATIONS] = calloc(1, sizeof(double[NO_OF_CHOSEN_OBSERVATIONS][NO_OF_CHOSEN_OBSERVATIONS]));
	/*
	downward sweep
	--------------
	to_be_inverted will become the r_matrix now (misuse of name)
	*/
	for (int i = 0; i < NO_OF_CHOSEN_OBSERVATIONS - 1; ++i)
	{
		l_matrix[i][i] = 1;
		for (int j = i + 1; j < NO_OF_CHOSEN_OBSERVATIONS; ++j)
		{
			l_matrix[j][i] = to_be_inverted[j][i]/to_be_inverted[i][i];
			for (int k = i; k < NO_OF_CHOSEN_OBSERVATIONS; ++k)
			{
				to_be_inverted[j][k] = to_be_inverted[j][k] - l_matrix[j][i]*to_be_inverted[i][k];
			}
		}
	}
	
	/*
	The LU decomposition is already done at this point.
	We now use the LU decomposition for the inversion.
	We know LU = A. We want to solve AA^-1 = 1, a.k.a. LUA^-1 = 1.
	Therefore, we firstly solve LB = 1 with a downward sweep.
	*/
	double (*b_matrix)[NO_OF_CHOSEN_OBSERVATIONS] = calloc(1, sizeof(double[NO_OF_CHOSEN_OBSERVATIONS][NO_OF_CHOSEN_OBSERVATIONS]));
	for (int i = 0; i < NO_OF_CHOSEN_OBSERVATIONS - 1; ++i)
	{
		b_matrix[i][i] = 1/l_matrix[i][i];
		for (int j = 0; j < i; ++j)
		{
			for (int k = 0; k < i; ++k)
			{
				b_matrix[i][j] -= l_matrix[i][k]*b_matrix[k][j]/l_matrix[i][i];
			}
		}
	}
	// l_matrix is not needed anymore
	free(l_matrix);
	
	/*
	Now we have to solve UA^-1 = B with an upward sweep.
	U is to_be_inverted (see above).
	*/
	for (int i = 0; i < NO_OF_CHOSEN_OBSERVATIONS; ++i)
	{
		inv[NO_OF_CHOSEN_OBSERVATIONS - 1][i] = b_matrix[NO_OF_CHOSEN_OBSERVATIONS - 1][i]/to_be_inverted[NO_OF_CHOSEN_OBSERVATIONS - 1][NO_OF_CHOSEN_OBSERVATIONS - 1];
		for (int j = NO_OF_CHOSEN_OBSERVATIONS - 2; j >= 0; --j)
		{
			for (int k = j + 1; k < NO_OF_CHOSEN_OBSERVATIONS ; ++j)
			{
				inv[j][i] = b_matrix[j][i] - to_be_inverted[j][k]*inv[k][i]/to_be_inverted[j][j];
			}
		}
	}
	
	// that's it, b_matrix is not needed anymore
	free(b_matrix);
	return 0;
}

File: src/lin_algebra.f90
! This source file is part of GAME-DA, which is released under the MIT license.
! Github repository: https://github.com/OpenNWP/GAME-DA

! This module contains linear algebra functions.

module lin_algebra

  use iso_c_binding

  implicit none

  ! precision of doubles
  integer, parameter        :: wp = c_double
  
  contains
    
  subroutine gauss(to_be_inverted, inv, matrix_size) &
  bind(c, name = "gauss")

    ! This subroutine computes the inverse inv of the matrix to_be_inverted, using the Gauss scheme.
    ! CAUTION: in the process, to_be_inverted will be modified.
    
    real(wp), intent(inout) :: to_be_inverted(matrix_size,matrix_size)
    real(wp), intent(inout) :: inv           (matrix_size,matrix_size)
    integer,  intent(in)    :: matrix_size

    ! local variables
    integer                 :: permute_index_found, permute_index_counter
    real(wp)                :: factor
    ! loop indices
    integer                 :: i, j, k

    ! firstly, the inverse is initialized with the unity matrix
    !$omp parallel
    !$omp do private(i)
    do i = 1,matrix_size
      inv(i,i) = 1._wp
    enddo
    !$omp end do
    !$omp end parallel

    ! Gaussian downwards
    ! ------------------
    ! we will start to modify to_be_inverted now (misuse of name)
    do i = 1,matrix_size-1
      ! checking if a permutation is necessary
      ! Firstly, the permutation index has to be found.
      permute_index_found = 0
      permute_index_counter = i
      do while (permute_index_found == 0)
        if (to_be_inverted(permute_index_counter,i) /= 0._wp) then
          permute_index_found = 1
        else
          permute_index_counter = permute_index_counter + 1
        endif
      enddo
      ! actually performing the permutation
      if (permute_index_counter > i) then
        call permute_lines(to_be_inverted, i, permute_index_counter, matrix_size)
        call permute_lines(inv, i, permute_index_counter, matrix_size)
      endif

      ! permutation is done, now comes the actual calculation
      ! dividing the line by to_be_inverted(i,i)
      factor = 1._wp/to_be_inverted(i,i)
      !$omp parallel
      !$omp do private(j)
      do j=i,matrix_size
        to_be_inverted(i,j) = factor*to_be_inverted(i,j)
      enddo
      !$omp end do
      !$omp end parallel
      !$omp parallel
      !$omp do private(j)
      do j = 1,matrix_size
        inv(i,j) = factor*inv(i,j)
      enddo
      !$omp end do
      !$omp end parallel
      ! loop over all the lines that are below the current line
      !$omp parallel
      !$omp do private(factor,j,k)
      do j=i+1,matrix_size
        factor = -to_be_inverted(j,i)
        do k=i,matrix_size
          to_be_inverted(j,k) = to_be_inverted(j,k) + factor*to_be_inverted(i,k)
        enddo
        inv(j,:) = inv(j,:) + factor*inv(i,:)
      enddo
      !$omp end do
      !$omp end parallel
    enddo

    !$omp parallel
    !$omp do private(j)
    do j = 1,matrix_size
      inv(matrix_size,j) = inv(matrix_size,j)/to_be_inverted(matrix_size,matrix_size)
    enddo
    !$omp end do
    !$omp end parallel
    to_be_inverted(matrix_size,matrix_size) = 1._wp

    ! Gaussian upwards
    ! ----------------
    do i = matrix_size,2,-1
      !$omp parallel
      !$omp do private(j)
      do j = i-1,1,-1
        inv(j,:) = inv(j,:) - to_be_inverted(j,i)*inv(i,:)
      enddo
      !$omp end do
      !$omp end parallel
    enddo
  
  end subroutine gauss

  subroutine permute_lines(matrix, line_a, line_b, matrix_size)

    ! This subroutine permutes line_a with line_b of matrix.

    ! arguments
    real(wp), intent(inout) :: matrix(matrix_size,matrix_size)
    integer,  intent(in)    :: line_a
    integer,  intent(in)    :: line_b
    integer,  intent(in)    :: matrix_size

    ! local variables
    real(wp)                :: line_a_pre(matrix_size)

    line_a_pre(:) = matrix(line_a,:)
    matrix(line_a,:) = matrix(line_b,:)
    matrix(line_b,:) = line_a_pre(:)

  end subroutine permute_lines

end module lin_algebra








